sources/osgShader/Compiler.js

```diff
@@ -947,10 +964,23 @@ Compiler.prototype = {
         }
         functor.call( functor, node );
 
-        // keep trace we visited.
+        // keep trace we visited
         this.markNodeAsVisited( node );
+
     },
 
+    // clean necessary bits before traversing
+    // called in each evaluate func belows
+    preTraverse: function ( visitor ) {
```

Comment:
I started with that, but was afraid "init" was implying a "once and for all"
Whereas "Pre" goes well with compiler language (multi-pass, pre-pass, post-pass) ?